Refactored WP_Debug_Data::debug_data() - get_wp_server()

apermo opened the pull request:
Changed: Moving $info['wp-server'] to own function

---

Follow up to https://github.com/WordPress/wordpress-develop/pull/7027 as discussed with @dmsnell and @costdev 

Summary of the coming steps:

1. [wp-filesystem](https://github.com/WordPress/wordpress-develop/pull/7065)
2. [wp-constants](https://github.com/WordPress/wordpress-develop/pull/7106)
3. [wp-database](https://github.com/WordPress/wordpress-develop/pull/7143)
4. **wp-server** (we are here)
5. wp-media
6. wp-plugins-active & wp-plugins-inactive
7. [wp-mu-plugins](https://github.com/WordPress/wordpress-develop/pull/7305)
8. wp-themes-inactive
9. wp-parent-theme
10. wp-active-theme
11. wp-dropins
12. wp-paths-sizes
13. wp-core

Refactoring WP_Debug_Data::debug_data(); into smaller functions for

* better maintainability
* reduced complexity
* preparation for future unit tests
* improved extensibility (see trac for details)
* added php7.2 style type hints
* fixed an error with `$parent_theme_auto_update_string` showing the value from `$auto_updates_string` (This will be handled only in Step 9)

Trac ticket: https://core.trac.wordpress.org/ticket/61648#ticket

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

apermo committed:
Refactoring Part 4 - wp-server

Moved wp-server to a distinct function.
Sorted function calls for wp-database

apermo commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1223,6 +1038,196 @@ public static function debug_data() {
+		}
+
+		// Server time.
+		$date = new DateTime( 'now', new DateTimeZone( 'UTC' ) );
```
@dmsnell PHPstorm warned me that this line could possibly throw an exception. Since we have a highly controlled set of parameters, I think it's safe to assume we will never get this exception.

Yet PHPstorm wanted me to add a `@throws` in the PHPdoc, which I ignored for the time being. But what do you suggest what we do about it?

Adding a `@throws` will just bubble up the problem. 

How about
```suggestion
		try {
			$date = new DateTime( 'now', new DateTimeZone( 'UTC' ) );
		} catch ( Exception $e ) {
			// It should be impossible to reach this statement. But in case let's fail gracefully.
			wp_die( $e->get_message() );
		}
```

(Untested and written in GitHub only)

costdev replied:
While not impossible (See the [Europe/Kyiv change](https://mm.icann.org/pipermail/tz-announce/2022-August/000071.html) in the "Changes to zone name" section - Which changed for PHP 8.2), since we're using the UTC timezone, the chance of an Exception being thrown for this call is extremely low.

It seems to be fairly safe to consider this a tooling flag that can be ignored.

I don't have strong feelings on this though. It'll probably just come down to preference.

dmsnell replied:
I don't see any notes in the documentation suggesting that `UTC` was ever _not_ part of the distribution. The constant `DateTimeZone::UTC` hints at this, and if you ask PHP to get a timezone supported by UTC it returns the string in use here.

```php
php > var_dump( DateTimeZone::listIdentifiers( DateTimeZone::UTC ) );
array(1) {
  [0]=>
  string(3) "UTC"
}
```

While sometimes the documentation is wrong, it seems unlikely here, meaning that it may not even be an extremely low chance of throwing here, but not possible.

Do we need a WPCS ignore statement here? If not then I think it's fine as-is. Regardless, as I have written in other comments - this PR is about modularizing a piece of the debug data code, and I recommend against making semantic changes otherwise until that factoring has completed.

apermo replied:
It does not trigger WPCS, just triggers the Code analysis from PHPstorm.

So we can keep it as is.

apermo commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1154,8 +968,9 @@ public static function debug_data() {
 		}
 
-		$info['wp-constants']  = self::get_wp_constants();
+		$info['wp-server']     = self::get_wp_server();
 		$info['wp-database']   = self::get_wp_database();
+		$info['wp-constants']  = self::get_wp_constants();
 		$info['wp-filesystem'] = self::get_wp_filesystem();
```
I reordered this, to be able to get rid of the `$info['wp-database']` in lines 199-202

Actually I would like to order the function calls bellow as well. OCD is itching.

This is the correct order of the function calls, and by having them correct, we can completely replace the original `$info` definition which was the plan of this.
And I'd like it best if we could still swap around the function defintions of `get_wp_database()` in line 1391 and `get_wp_constants()` in line 1238 to match this order.

Your thought @dmsnell ?

dmsnell replied:
After discussing in Slack, this all sounds great, and I apologize for reordering these before. The insertion order of the keys in the array determines the order they are displayed in `wp-admin`, and so we want to preserve the existing order.

dmsnell replied:
@apermo thoughts on this? when I dug into it, I felt like something ad-hoc was missing a bigger point, so I proposed encoding the ordering as a specific semantic instead of relying on how the code is already written.

https://github.com/WordPress/wordpress-develop/pull/7289

apermo replied:
@dmsnell updated the PR.

I'll add an inline comment and tag you, you've missed one thing in #7289 that we need to cover in order to match the format for both single and multisite.

mukeshpanchal27 commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1223,6 +1038,196 @@ public static function debug_data() {
+	 *
+	 * @return array
+	 */
+	public static function get_wp_server(): array {
```
```suggestion
	public static function get_wp_server() {
```

 I don't think we need strict typing on return values.

apermo replied:
All previously added functions have a return type, so I'd say we should stick to it and have the return type here.

kebbet commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1223,6 +1038,196 @@ public static function debug_data() {
+		// Whether PHP supports 64-bit.
+		$php64bit = ( PHP_INT_SIZE * 8 === 64 );
+
+		$php_version = sprintf(
+			'%s %s',
+			$php_version_debug,
+			( $php64bit ? __( '(Supports 64bit values)' ) : __( '(Does not support 64bit values)' ) )
+		);
```
Could be turned into 
```
 if ( $php64bit ) {
     $php_version = sprintf(
         /* translators: PHP version number. */
         _( '%s (Supports 64bit values)' ),
         $php_version_debug
     );
} else {
     $php_version = sprintf(
         /* translators: PHP version number. */
         _( '%s (Does not support 64bit values)' ),
         $php_version_debug
     );
}
```

dmsnell replied:
given the scope of changes in this sequence of PRs, I think it would be ideal if we stuck to a modularization of the method first and then followed-up with "improvements" afterwards. these are great ideas to earmark once the changes are complete.

kebbet replied:
Sure, I understand! Just as a note though: one, possibly two, strings were adjusted in the wp-database PR. I can create a follow up ticket to the main ticket that focus on improvements, such as reviewing strings.

kebbet commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1223,6 +1038,196 @@ public static function debug_data() {
+			);
+		} else {
+			$fields['max_input_variables'] = array(
+				'label' => __( 'PHP max input variables' ),
```
Could this label be written in a more user friendly way?
`Max number of PHP input variables` ?

dmsnell replied:
same comment as earlier: this is a good thing to return to after all of the methods have been created and the entire thing split up.

kebbet commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1223,6 +1038,196 @@ public static function debug_data() {
+			}
+
+			$fields['max_input_time']      = array(
+				'label' => __( 'Max input time' ),
```
Needs clarification?

apermo replied:
As @dmsnell said, this is primarily about refactoring the existing code into functions. It's just copy and paste of the original code, and just doing search replace `$info['wp-server']['fields']` => `$fields`

I would ommit any further changes, since we want to finish another 8 or 9 PR before the release of 6.7. if possible.

kebbet commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1223,6 +1038,196 @@ public static function debug_data() {
+				'value' => ini_get( 'max_input_time' ),
+			);
+			$fields['upload_max_filesize'] = array(
+				'label' => __( 'Upload max filesize' ),
```
Need clarification too?

apermo replied:
same as above.

kebbet reviewed (commented):
Strings could be more clear, while we are editing this function as a whole.

apermo committed:
Refactored WP_Debug_Data::debug_data() Part 4

Adaptations for #7289
Added a condition to unset $info['wp-paths-sizes'] for multisites to match the original state.

apermo commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -85,12 +85,17 @@ public static function debug_data() {
 
+		// Remove accordion for Directories and Sizes if in Multisite.
+		if ( $is_multisite ) {
+			unset( $info['wp-paths-sizes'] );
```
@dmsnell See lines 167-173 and 288-348, `$info['wp-paths-sizes']` shall only be visible on single sites, so I've added this unset here. I think that was why I originally opted for `$info['key'] = self::get_key;` instead of your approach. 

And I've knowingly used `is_multisite()` instead of `$is_multisite`, we will remove `$is_multisite = is_multisite();`(Line 44) later in the refactoring, since it is likely only helpful in the upcoming `function get_wp_core()`

dmsnell replied:
this seemms fine. thanks @apermo

dmsnell replied:
I've updated the comment in f576b043ae

dmsnell replied:
we could eventually filter this array and have `self::get_wp_paths_sizes()` return `null` for multi-site deployments.

apermo committed:
Refactored WP_Debug_Data::debug_data() Part 4

Changed: replaced $is_multisite with is_multisite() as preparation for a later step.

dmsnell commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -85,12 +85,17 @@ public static function debug_data() {
 			'wp-filesystem'       => self::get_wp_filesystem(),
 		);
 
+		// Remove accordion for Directories and Sizes if in Multisite.
```
This could probably use some rearranging. I understand that the display of the debug data in `wp-admin` happens to display sections in a UI like an accordion, but this function doesn't directly couple with that, does it?

Could the comment focus on the purpose of the debug data itself, and not the display?

```php
// Path sizes are only relevant in single-site contexts.
```

dmsnell committed:
Update comment removing pre-filled paths-sizes data.

dmsnell commented on src/wp-admin/includes/class-wp-debug-data.php:
```diff
@@ -1246,6 +1065,196 @@ public static function debug_data() {
+			$php_version_debug .= ' 64bit';
+		}
+
+		$fields['server_architecture'] = array(
```
we seem to be assigning keys to `$fields` here without initializing `$fields` as a variable.

dmsnell replied:
initialized in e6f03eaa28

dmsnell committed:
Initialize $fields before assignment.

dmsnell committed:
Merge remote-tracking branch 'upstream' into refactor-wp-debug-data/part4

dmsnell commented:
Merged in [[59002]](https://core.trac.wordpress.org/changeset/59002)
https://github.com/wordpress/wordpress-develop/commit/19fc9c7aadcaf24b51b3479dd62121b74213a7d7
